src/debug.h

```diff
@@ -5,38 +5,47 @@
 #define _DEBUG_H
 
 /* Debugging */
-extern int debug_lazy_init;
-#define P11PROV_debug_status(action) \
+extern int debug_level;
+#define P11PROV_debug(...) \
     do { \
-        int enabled = 0; \
-        if (__atomic_compare_exchange_n(&debug_lazy_init, &enabled, -1, true, \
-                                        __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)) { \
+        if (debug_level < 0) { \
             p11prov_debug_init(); \
-        } \
-        if (enabled >= 1) { \
-            action; \
+        } else if (debug_level > 0) { \
+            p11prov_debug(OPENSSL_FILE, OPENSSL_LINE, OPENSSL_FUNC, \
+                          __VA_ARGS__); \
         } \
     } while (0)
 
-#define P11PROV_debug(...) P11PROV_debug_status(p11prov_debug(__VA_ARGS__))
-
 #define P11PROV_debug_mechanism(...) \
-    P11PROV_debug_status(p11prov_debug_mechanism(__VA_ARGS__))
+    do { \
+        if (debug_level < 0) { \
+            p11prov_debug_init(); \
+        } else if (debug_level > 0) { \
+            p11prov_debug_mechanism(__VA_ARGS__); \
+        } \
+    } while (0)
 
 #define P11PROV_debug_slot(...) \
-    P11PROV_debug_status(p11prov_debug_slot(__VA_ARGS__))
+    do { \
+        if (debug_level < 0) { \
+            p11prov_debug_init(); \
+        } else if (debug_level > 0) { \
+            p11prov_debug_slot(__VA_ARGS__); \
```

Comment:
Should not this be separate `if` (similarly in the other macros here)? Isn't this going to ignore the first message and do just the initialization?

Reply:
good catch!

Reply:
Fixed in rebased patch